File: vaidcord-go/guild.go
package vaidcord

type Guild struct {
	ID                          string   `json:"id"`
	Name                        string   `json:"name"`
	Icon                        string   `json:"icon,omitempty"`
	Splash                      string   `json:"splash,omitempty"`
	Description                 string   `json:"description,omitempty"`
	OwnerID                     string   `json:"owner_id,omitempty"`
	AFKChannelID                string   `json:"afk_channel_id,omitempty"`
	AFKTimeout                  int      `json:"afk_timeout,omitempty"`
	VerificationLevel           int      `json:"verification_level,omitempty"`
	DefaultMessageNotifications int      `json:"default_message_notifications,omitempty"`
	ExplicitContentFilter       int      `json:"explicit_content_filter,omitempty"`
	Roles                       []Role   `json:"roles,omitempty"`
	Features                    []string `json:"features,omitempty"`
	MFALevel                    int      `json:"mfa_level,omitempty"`
	ApplicationID               string   `json:"application_id,omitempty"`
	SystemChannelID             string   `json:"system_channel_id,omitempty"`
	RulesChannelID              string   `json:"rules_channel_id,omitempty"`
	MaxMembers                  int      `json:"max_members,omitempty"`
	VanityURLCode               string   `json:"vanity_url_code,omitempty"`
	Banner                      string   `json:"banner,omitempty"`
	PremiumTier                 int      `json:"premium_tier,omitempty"`
	PreferredLocale             string   `json:"preferred_locale,omitempty"`
	NSFWLevel                   int      `json:"nsfw_level,omitempty"`
	// Extra fields present on the GUILD_CREATE gateway payload.
	JoinedAt    string    `json:"joined_at,omitempty"`
	Large       bool      `json:"large,omitempty"`
	Unavailable bool      `json:"unavailable,omitempty"`
	MemberCount int       `json:"member_count,omitempty"`
	Members     []Member  `json:"members,omitempty"`
	Channels    []Channel `json:"channels,omitempty"`
	Threads     []Channel `json:"threads,omitempty"`
}

type Role struct {
	ID           string    `json:"id"`
	Name         string    `json:"name"`
	Color        int       `json:"color,omitempty"`
	Hoist        bool      `json:"hoist,omitempty"`
	Icon         string    `json:"icon,omitempty"`
	UnicodeEmoji string    `json:"unicode_emoji,omitempty"`
	Position     int       `json:"position,omitempty"`
	Permissions  string    `json:"permissions,omitempty"`
	Managed      bool      `json:"managed,omitempty"`
	Mentionable  bool      `json:"mentionable,omitempty"`
	Tags         *RoleTags `json:"tags,omitempty"`
	Flags        int       `json:"flags,omitempty"`
}

type RoleTags struct {
	BotID         string `json:"bot_id,omitempty"`
	IntegrationID string `json:"integration_id,omitempty"`
}

type Member struct {
	User                       *User    `json:"user,omitempty"`
	Nick                       string   `json:"nick,omitempty"`
	Avatar                     string   `json:"avatar,omitempty"`
	Roles                      []string `json:"roles,omitempty"`
	JoinedAt                   string   `json:"joined_at,omitempty"`
	PremiumSince               string   `json:"premium_since,omitempty"`
	Deaf                       bool     `json:"deaf,omitempty"`
	Mute                       bool     `json:"mute,omitempty"`
	Flags                      int      `json:"flags,omitempty"`
	Pending                    bool     `json:"pending,omitempty"`
	Permissions                string   `json:"permissions,omitempty"`
	CommunicationDisabledUntil string   `json:"communication_disabled_until,omitempty"`
}

File: vaidcord-rust/src/models/guild.rs
//! Guild, role and member models.

use serde::Deserialize;
use serde_json::Value;

use super::User;

/// A Discord guild ("server").
///
/// Unknown fields in the payload are ignored so the model stays tolerant of
/// API additions.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Default)]
pub struct Guild {
    pub id: String,
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub icon: Option<String>,
    #[serde(default)]
    pub owner_id: Option<String>,
    #[serde(default)]
    pub afk_channel_id: Option<String>,
    #[serde(default)]
    pub afk_timeout: Option<u64>,
    #[serde(default)]
    pub verification_level: Option<u8>,
    #[serde(default)]
    pub roles: Vec<Role>,
    #[serde(default)]
    pub emojis: Vec<Value>,
    #[serde(default)]
    pub features: Vec<String>,
    #[serde(default)]
    pub member_count: Option<u64>,
    #[serde(default)]
    pub members: Vec<Member>,
    #[serde(default)]
    pub channels: Vec<Value>,
    #[serde(default)]
    pub threads: Vec<Value>,
    #[serde(default)]
    pub joined_at: Option<String>,
    #[serde(default)]
    pub large: Option<bool>,
    #[serde(default)]
    pub unavailable: Option<bool>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub preferred_locale: Option<String>,
    #[serde(default)]
    pub system_channel_id: Option<String>,
    #[serde(default)]
    pub rules_channel_id: Option<String>,
    #[serde(default)]
    pub premium_tier: Option<u8>,
    #[serde(default)]
    pub nsfw_level: Option<u8>,
}

/// A guild role.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Default)]
pub struct Role {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub color: Option<u32>,
    #[serde(default)]
    pub hoist: bool,
    #[serde(default)]
    pub icon: Option<String>,
    #[serde(default)]
    pub unicode_emoji: Option<String>,
    #[serde(default)]
    pub position: Option<i64>,
    #[serde(default)]
    pub permissions: Option<String>,
    #[serde(default)]
    pub managed: bool,
    #[serde(default)]
    pub mentionable: bool,
    #[serde(default)]
    pub flags: Option<u64>,
}

/// A guild member.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Default)]
pub struct Member {
    #[serde(default)]
    pub user: Option<User>,
    #[serde(default)]
    pub nick: Option<String>,
    #[serde(default)]
    pub avatar: Option<String>,
    #[serde(default)]
    pub roles: Vec<String>,
    #[serde(default)]
    pub joined_at: Option<String>,
    #[serde(default)]
    pub premium_since: Option<String>,
    #[serde(default)]
    pub deaf: bool,
    #[serde(default)]
    pub mute: bool,
    #[serde(default)]
    pub pending: Option<bool>,
    #[serde(default)]
    pub permissions: Option<String>,
    #[serde(default)]
    pub communication_disabled_until: Option<String>,
    #[serde(default)]
    pub flags: Option<u64>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn guild_decodes_with_unknown_fields() {
        let guild: Guild = serde_json::from_str(
            r#"{"id":"1","name":"Test","owner_id":"42","roles":[{"id":"9","name":"admin","permissions":"8"}],"some_future_field":{"x":1}}"#,
        )
        .unwrap();
        assert_eq!(guild.id, "1");
        assert_eq!(guild.name.as_deref(), Some("Test"));
        assert_eq!(guild.roles.len(), 1);
        assert_eq!(guild.roles[0].permissions.as_deref(), Some("8"));
    }

    #[test]
    fn member_decodes_with_partial_payload() {
        let member: Member =
            serde_json::from_str(r#"{"roles":["1","2"],"nick":"vai","mute":true}"#).unwrap();
        assert_eq!(member.roles.len(), 2);
        assert!(member.mute);
        assert!(member.user.is_none());
    }
}
